File: leetcode/scramble_string/ScrambleString.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        if (s1.size() != s2.size()) return false; 
        if (s1 == s2) return true; 
        
        string temp1 = s1; 
        string temp2 = s2; 
        sort(temp1.begin(), temp1.end()); 
        sort(temp2.begin(), temp2.end()); 
        
        if (temp1 != temp2) return false; 
        
        // Try all split
        int size = s1.size(); 
        
        for (int i = 1; i < size; i++) {
        	string s1l = s1.substr(0, i); 
        	string s1r = s1.substr(i, size - i); 
        	
        	// Do not swap s2
        	string s2l = s2.substr(0, i); 
        	string s2r = s2.substr(i, size - i); 
        	
        	if (isScramble(s1l, s2l) && isScramble(s1r, s2r)) return true; 
        	
        	// Swap s2
        	s2l = s2.substr(0, size - i); 
        	s2r = s2.substr(size - i, i); 
        	
        	if (isScramble(s1l, s2r) && isScramble(s1r, s2l)) return true; 
        }
        
        return false; 
    }
};

File: leetcode4/ScrambleString.java
public class Solution {
    public boolean isScramble(String s1, String s2) {
        int len1 = s1.length(); 
        int len2 = s2.length(); 
        
        if (len1 != len2) {
            return false; 
        }
        
        if (len1 == 0) {
            return true; 
        }
        
        boolean[][][] T = new boolean[len1][len1][len1 + 1]; 
        
        for (int i = 0; i < len1; i++) {
            for (int j = 0; j < len1; j++) {
                if (s1.charAt(i) == s2.charAt(j)) {
                    T[i][j][1] = true; 
                } else {
                    T[i][j][1] = false; 
                }
            }
        }
        for (int k = 2; k <= len1; k++) {
            for (int i = 0; i < len1; i++) {
                for (int j = 0; j < len1; j++) {
                    if (i + k <= len1 && j + k <= len1) {
                        for (int x = 1; x < k; x++) {
                            T[i][j][k] |= T[i][j][x] && T[i + x][j + x][k - x];
                            T[i][j][k] |= T[i][j + k - x][x] && T[i + x][j][k - x]; 
                            
                            if (T[i][j][k]) {
                                break; 
                            }
                        }
                    }
                }
            }
        } 
        
        return T[0][0][len1]; 
    }
}
